staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go

```diff
@@ -515,6 +457,74 @@ func (cnc *CloudNodeController) getNodeModifiersFromCloudProvider(ctx context.Co
 	return nodeModifiers, nil
 }
 
+func (cnc *CloudNodeController) getProviderID(ctx context.Context, node *v1.Node) (string, error) {
+	if node.Spec.ProviderID != "" {
+		return node.Spec.ProviderID, nil
+	}
+
+	providerID, err := cloudprovider.GetInstanceProviderID(ctx, cnc.cloud, types.NodeName(node.Name))
+	if err == cloudprovider.NotImplemented {
+		// if the cloud provider being used does not support provider IDs,
+		// we can safely continue since we will attempt to set node
+		// addresses given the node name in getNodeAddressesByProviderIDOrName
+		klog.Warningf("cloud provider does not set node provider ID, using node name to discover node %s", node.Name)
+		return "", nil
+	}
+
+	// if the cloud provider being used supports provider IDs, we want
+	// to propagate the error so that we re-try in the future; if we
+	// do not, the taint will be removed, and this will not be retried
+	return providerID, err
+}
+
+// getInstanceMetadata get instance type and nodeAddresses, use Instances if InstancesV2 is off.
+func (cnc *CloudNodeController) getInstanceMetadata(ctx context.Context, node *v1.Node) (*cloudprovider.InstanceMetadata, error) {
```

Comment:
@andrewsykim  one concern here: when providerID not passed here, `InstanceMetadata`, `getNodeAddressesByProviderIDOrName` and `getInstanceTypeByProviderIDOrName` could still work fine, right?

Reply:
> @andrewsykim one concern here

No more since I set ProviderID in line 474.